### PR TITLE
feat: Add CG, TN, and line search algorithm that will be used in OFDFT.

### PR DESCRIPTION
**Pull request by sunliang98**
Add CG method, Truncated-Newton, and line search algorithm and corresponding unittests to source/module_esolver/opt/, these methods will be used in OFDFT later.

### CODE REVIEW

File: source/module_esolver/opt/opt_DCsrch.cpp [hunk 1]
```diff
@@ -0,0 +1,750 @@
+#include "./opt_DCsrch.h"
+
+extern "C"{
+/* Copyright (C) 1996 Jorge J. More'.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
```
Comment (caic99) [inline source/module_esolver/opt/opt_DCsrch.cpp]: This source file is using a different license from ABACUS.

File: source/module_esolver/opt/test/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,21 @@
+cmake_minimum_required(VERSION 3.20)
+
+project(CG_test)
+project(TN_test)
```
Comment (caic99) [inline source/module_esolver/opt/test/CMakeLists.txt:4]: Why do we need a independent CMake project here?

Comment (caic99) [reply source/module_esolver/opt/test/CMakeLists.txt:4]: For tests, please refer to our [docs](https://github.com/deepmodeling/abacus-develop/blob/develop/docs/CONTRIBUTING.md#adding-a-unit-test).

Comment (sunliang98) [reply source/module_esolver/opt/test/CMakeLists.txt:4]: I see, CG and TN have not been used in ABACUS for now, so I add this CMake project to compile the files in /source/module_esolver/opt separately. I will remove it and modify the tests to standard unittest when OFDFT module is completed.

File: source/module_esolver/opt/opt_DCsrch.cpp (hunk 1 above)
Comment (sunliang98) [reply source/module_esolver/opt/opt_DCsrch.cpp]: I have removed this file temporarily. Later it will be replaced.

**Review by mohanchen — APPROVED** [review]
agree

**Pull request merged**